shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Scripts.hs

```diff
@@ -0,0 +1,266 @@
+--   The key component is the class ScriptClass.
+module Test.Shelley.Spec.Ledger.Generator.Scripts
+  ( ScriptClass(..),
+    Quantifier(..),
+    anyOf, allOf, mOf,
+    ScriptPairs,
+    keyPairs,
+    mkPayScriptHashMap,
+    mkStakeScriptHashMap,
+    mkScriptsFromKeyPair,
+    mkKeyPairs,
+    mkScripts,
+    mkScriptCombinations,
+    combinedScripts,
+    someScripts,
+    scriptKeyCombinations,
+    scriptKeyCombination,
+  )
+  where
+
+import Shelley.Spec.Ledger.Scripts(MultiSig(..),ScriptHash,hashMultiSigScript)
+import Cardano.Ledger.Shelley (ShelleyEra)
+import qualified Cardano.Ledger.Core as Core
+import Cardano.Ledger.Core(Script,Value,TxBody,ChainData, SerialisableData, AnnotatedData)
+import Cardano.Ledger.Era (Era(..))
+import qualified Cardano.Ledger.Crypto as CC (Crypto)
+import Shelley.Spec.Ledger.Keys (Hash, KeyHash, KeyRole (..),VKey(..), hashKey, asWitness, KeyPair(..), vKey)
+import Cardano.Binary(ToCBOR(..),FromCBOR(..),Annotator)
+import Cardano.Ledger.Torsor (Torsor (..))
+import Cardano.Ledger.Compactible(Compactible(..),CompactForm(..))
+import NoThunks.Class (NoThunks (..))
+import Shelley.Spec.Ledger.Hashing (EraIndependentTxBody, HashAnnotated (..))
+import Data.Proxy
+import qualified Data.Map as Map
+import qualified Data.List as List
+import Cardano.Ledger.Val(Val(..))
+import qualified Cardano.Crypto.Hash as Hash
+import Shelley.Spec.Ledger.TxBody() -- import instances only
+import Data.List(concatMap,permutations)
+import Shelley.Spec.Ledger.LedgerState( KeyPairs )
+import Test.Shelley.Spec.Ledger.Generator.Constants
+  ( Constants (..),
+    defaultConstants,
+  )
+import Test.QuickCheck (Gen)
+import qualified Test.QuickCheck as QC
+import Test.Shelley.Spec.Ledger.Utils( mkKeyPair )
+import Data.Word(Word64)
+import Cardano.Crypto.DSIGN.Class (DSIGNAlgorithm (..))
+import Cardano.Ledger.Crypto (DSIGN)
+import Data.Tuple (swap)
+import Shelley.Spec.Ledger.Tx( ValidateScript (..) )
+
+
+-- ==============================================================================
+
+class ( Show (Script era),
+        Eq (Script era),
+        ValidateScript era,
+        HashIndex (Core.TxBody era) ~ EraIndependentTxBody,
+        Eq (Core.TxBody era),
+        Show (Core.TxBody era),
+        CC.Crypto(Crypto era),
+        Era era ) => ScriptClass era where
+  basescript :: Proxy era -> KeyHash 'Witness (Crypto era) -> Script era
+  isKey :: Proxy era -> Script era -> Maybe (KeyHash 'Witness (Crypto era))
+  quantify :: Proxy era -> Script era -> Quantifier (Script era)
+  unQuantify:: Proxy era -> Quantifier (Script era) -> Script era
+
+type ScriptPairs era = [(Core.Script era, Core.Script era)]
+
+-- ====================================
+
+-- | Abstracts the quantifier structure of (Script era)
+--   used in the 'quantify' and 'unQuantify' methods of ScriptClass.
+data Quantifier t = AllOf [t] | AnyOf [t] | MOf Int [t] | Leaf t
+
+anyOf :: forall era. ScriptClass era => Proxy era -> [Script era] -> Script era
```

Comment:
Given that we already have `AllowAmbiguousTypes` enabled, is there much point in having the proxy here? We can just resolve this with type annotations, without the extra argument.